esphome/components/dfplayer/__init__.py

```diff
@@ -0,0 +1,205 @@
+import esphome.codegen as cg
+import esphome.config_validation as cv
+from esphome import automation
+from esphome.const import CONF_ID, CONF_TRIGGER_ID
+from esphome.components import uart
+
+DEPENDENCIES = ['uart']
+
+dfplayer_ns = cg.esphome_ns.namespace('dfplayer')
+DFPlayer = dfplayer_ns.class_('DFPlayer', cg.Component)
+DFPlayerFinishedPlaybackTrigger = dfplayer_ns.class_('DFPlayerFinishedPlaybackTrigger',
+                                                     automation.Trigger.template())
+DFPlayerIsPlayingCondition = dfplayer_ns.class_('DFPlayerIsPlayingCondition', automation.Condition)
+
+MULTI_CONF = True
+CONF_FOLDER = 'folder'
+CONF_FILE = 'file'
+CONF_LOOP = 'loop'
+CONF_VOLUME = 'volume'
+CONF_EQ_PRESET = 'eq_preset'
+CONF_ON_FINISHED_PLAYBACK = 'on_finished_playback'
+
+EqPreset = dfplayer_ns.enum("EqPreset")
+EQ_PRESET = {
+    'NORMAL': EqPreset.NORMAL,
+    'POP': EqPreset.POP,
+    'ROCK': EqPreset.ROCK,
+    'JAZZ': EqPreset.JAZZ,
+    'CLASSIC': EqPreset.CLASSIC,
+    'BASS': EqPreset.BASS,
+}
+
+NextAction = dfplayer_ns.class_('NextAction', automation.Action)
+PreviousAction = dfplayer_ns.class_('PreviousAction', automation.Action)
+PlayFileAction = dfplayer_ns.class_('PlayFileAction', automation.Action)
+PlayFolderAction = dfplayer_ns.class_('PlayFolderAction', automation.Action)
+SetVolumeAction = dfplayer_ns.class_('SetVolumeAction', automation.Action)
+SetEqAction = dfplayer_ns.class_('SetEqAction', automation.Action)
+SleepAction = dfplayer_ns.class_('SleepAction', automation.Action)
+ResetAction = dfplayer_ns.class_('ResetAction', automation.Action)
+StartAction = dfplayer_ns.class_('StartAction', automation.Action)
+PauseAction = dfplayer_ns.class_('PauseAction', automation.Action)
+StopAction = dfplayer_ns.class_('StopAction', automation.Action)
+RandomAction = dfplayer_ns.class_('RandomAction', automation.Action)
+
+CONFIG_SCHEMA = cv.All(cv.Schema({
+    cv.GenerateID(): cv.declare_id(DFPlayer),
+    cv.Optional(CONF_ON_FINISHED_PLAYBACK): automation.validate_automation({
+        cv.GenerateID(CONF_TRIGGER_ID): cv.declare_id(DFPlayerFinishedPlaybackTrigger),
+    }),
+}).extend(uart.UART_DEVICE_SCHEMA))
+
+
+def to_code(config):
+    var = cg.new_Pvariable(config[CONF_ID])
+    yield cg.register_component(var, config)
+    yield uart.register_uart_device(var, config)
+
+    for conf in config.get(CONF_ON_FINISHED_PLAYBACK, []):
+        trigger = cg.new_Pvariable(conf[CONF_TRIGGER_ID], var)
+        yield automation.build_automation(trigger, [], conf)
+
+
+@automation.register_action('dfplayer.next', NextAction, cv.Schema({
```

Comment:
```suggestion
@automation.register_action('dfplayer.play_next', NextAction, cv.Schema({
```

Not sure about this suggestion - personally I think being more explicit about what `next` means is better, but if the next button also does other things then just `next` is also fine.

Reply:
It's ok for me, I went back and forward several times on the "design" but this device functions are not very consistent so I ended up doing the implementation of most functions just 1:1 as they are in the docs. But renaming this is ok